clicommand/git_credentials_helper.go

```diff
@@ -0,0 +1,134 @@
+)
+
+type GitCredentialsHelperConfig struct {
+	JobID string `cli:"job-id" validate:"required"`
+
+	// Global flags
+	Debug       bool     `cli:"debug"`
+	LogLevel    string   `cli:"log-level"`
+	NoColor     bool     `cli:"no-color"`
+	Experiments []string `cli:"experiment" normalize:"list"`
+	Profile     string   `cli:"profile"`
+
+	// API config
+	DebugHTTP        bool   `cli:"debug-http"`
+	AgentAccessToken string `cli:"agent-access-token" validate:"required"`
+	Endpoint         string `cli:"endpoint" validate:"required"`
+	NoHTTP2          bool   `cli:"no-http2"`
+}
+
+var GitCredentialsHelperCommand = cli.Command{
+	Name:  "git-credentials-helper",
+	Usage: "Ask buildkite for credentials to use to authenticate with Github when cloning. Uses the git-credential format, and is intended to be used as a git credential helper",
+	Flags: []cli.Flag{
+		cli.StringFlag{
+			Name:   "job-id",
+			Usage:  "The job id to get credentials for",
+			EnvVar: "BUILDKITE_JOB_ID",
+		},
+
+		// API Flags
+		AgentAccessTokenFlag,
+		EndpointFlag,
+		NoHTTP2Flag,
+		DebugHTTPFlag,
+
+		// Global flags
+		NoColorFlag,
+		DebugFlag,
+		LogLevelFlag,
+		ExperimentsFlag,
+		ProfileFlag,
+	},
+	Action: func(c *cli.Context) error {
+		ctx := context.Background()
+		ctx, cfg, l, _, done := setupLoggerAndConfig[GitCredentialsHelperConfig](ctx, c)
+		defer done()
+
+		l.Info("Authenticating checkout using Buildkite Github App Credentials...")
+
+		// ie, if the flags are from the command line rather than from the environment, which is how they should be passed
+		// to this process when it's called through the job executor
+		if os.Getenv("BUILDKITE_JOB_ID") == "" {
+			l.Warn("📎💬 It looks like you're calling this command directly in a step, rather than having the agent automatically call it")
+			l.Warn("This command is intended to be used as a git credential helper, and not called directly.")
+		}
+
+		// git passes the details of the current clone process to the credential helper via stdin
+		// we need to parse this to get the repo URL
+		// see: https://git-scm.com/docs/git-credential
+		stdin, err := io.ReadAll(os.Stdin)
+		if err != nil {
+			return fatal(c, l, fmt.Errorf("failed to read stdin: %v", err))
+		}
+
+		lines := strings.Split(string(stdin), "\n")
+
+		host := ""
+		path := ""
+		protocol := ""
+		for _, line := range lines {
+			if p, ok := strings.CutPrefix(line, "protocol="); ok {
+				protocol = strings.TrimSpace(p)
+			}
+			if p, ok := strings.CutPrefix(line, "host="); ok {
+				host = strings.TrimSpace(p)
+			}
+			if p, ok := strings.CutPrefix(line, "path="); ok {
+				path = strings.TrimSpace(p)
+			}
+		}
+
+		if protocol != "https" {
+			return fatal(c, l, fmt.Errorf("git remote must be using the https protocol to use Github App authentication"))
+		}
+
+		ep, err := url.JoinPath(host, path)
+		if err != nil {
+			return fatal(c, l, fmt.Errorf("failed to join host and path: %v", err))
+		}
+
+		repo := protocol + "://" + ep
```

Comment:
Let's factor out the parsing to make it testable. We should have some regression tests for this because I doubt we're going to use it for reals in the agent pipeline any time soon